Create default EE in seperate awx-managment cmd

kdelee opened the pull request:
Create EE at a seperate time and also attach a registry credential if
auth information provided

kdelee committed:
Create default EE in seperate awx-managment cmd

Create EE at a seperate time and also attach a registry credential if
auth information provided

This command can be run multiple times on the same instance and should be
idempotent.

jbradberry commented:
@kdelee do we have any associated PRs that make use of this new command during install or upgrade?

jbradberry commented:
This PR addresses the problem in #10287.  Also, #9851, but we need to change the packaging to invoke this command.

kdelee commented:
@jbradberry I've got  a PR for the downstream traditional installer, but we will need to make a PR to the operator to use this, except in the operator we may need to some extra finess around the secrets because we won't be able to directly pass secrets to this command

for operator:
 - need to call the register_default_execution_environment commands
 - for ee that need auth, we need to pull secret from an openshift secret, see use of k8s_info for other secrets like the SECRET_KEY
 - https://github.com/ansible/awx-operator/blob/devel/roles/installer/tasks/broadcast_websocket_configuration.yml#L10-L15

Zokormazo commented:
Since `DEFAULT_EXECUTION_ENVIRONMENTS` is a list, I went with 2 ees there and ended with two credentials named as `Default Execution Environment Registry Credential` . Since we are not allowing to set a credential per image, could we just reuse the first generated one?

kdelee commented:
@Zokormazo thats strange, did you call the command twice? I was hoping using `update_or_create` on the credential would prevent that. Let me try on my side again

kdelee commented:
@Zokormazo you were right :D I fixed my use of `update_or_create` and it should work now. Can you try it out?

kdelee commented on awx/main/management/commands/register_default_execution_environments.py:
```diff
@@ -0,0 +1,89 @@
+    Creates or updates the execution environments set in settings.DEFAULT_EXECUTION_ENVIRONMENTS if they are not yet created.
+    Optionally provide authentication details to create or update a container registry credential that will be set on all of these default execution environments.
+    Note that settings.DEFAULT_EXECUTION_ENVIRONMENTS is and ordered list, the first in the list will be used for project updates and system jobs.
+    """
```
@shanemcd @jbradberry I updated this help text, what do you think?

jbradberry replied:
"... is an ordered list"

kdelee committed:
Allow modifications of managed_by_tower creds

Sysadmins may need to modify managed_by_tower credentials, the
only known case of which is a default container registry credential
for use by the default execution environments, which are also managed_by_tower
and allow modifications.

kdelee commented on awx/api/serializers.py:
```diff
@@ -2603,8 +2603,6 @@ def get_summary_fields(self, obj):
 
     def validate(self, attrs):
-        if self.instance and self.instance.managed_by_tower:
-            raise PermissionDenied(detail=_("Modifications not allowed for managed credentials"))
```
@jbradberry mentioned I should revert this

kdelee committed:
revert change to serializer and use get_or_create

still need to fix up the handling of the verify-ssl bool, since
it is not being respected

shanemcd committed:
Tweaks to register_default_execution_environments command

- The credential will always be updated. I dont think we want to decrypt the password here for comparison
- Preserve newlines in help text

rebeccahhh reviewed (approved):
I like this a lot

shanemcd committed:
Register default EEs when booting dev env

kdelee commented on awx/main/management/commands/register_default_execution_environments.py:
```diff
@@ -0,0 +1,104 @@
+        )
+        parser.add_argument(
+            "--verify-ssl",
+            type=lambda x: bool(strtobool(str(x))),
```
nice
